File: JS/src/day02.js
module.exports.part1 = function (input) {
  return input.filter((s) => {
    const parts = s.split(" ");
    const min = parseInt(parts[0].split("-")[0]);
    const max = parseInt(parts[0].split("-")[1]);
    const letter = parts[1].slice(0, 1);
    const password = parts[2];

    const count = charCount(password, letter);

    return count >= min && count <= max;
  }).length;
};

function charCount(s, c) {
  let count = 0;
  let pos = s.indexOf(c);
  while (pos >= 0) {
    count++;
    pos = s.indexOf(c, pos + 1);
  }
  return count;
}

module.exports.part2 = function (input) {
  return input.filter((s) => {
    const parts = s.split(" ");
    const p1 = parseInt(parts[0].split("-")[0]);
    const p2 = parseInt(parts[0].split("-")[1]);
    const letter = parts[1].slice(0, 1);
    const password = parts[2];

    return (
      (password.slice(p1 - 1, p1) === letter) !==
      (password.slice(p2 - 1, p2) === letter)
    );
  }).length;
};

File: CSharpDotNetCore/AOC2020/Day02.cs
using System.Collections.Generic;
using System.Linq;

namespace AOC2020
{
    public static class Day02
    {
        public static int part1(IEnumerable<string> input) => validPasswords(input);

        private static int validPasswords(IEnumerable<string> input) => input.Where(isValid1).Count();

        private static bool isValid1(string input)
        {
            var parts = input.Split(" ");
            var min = int.Parse(parts[0].Split("-")[0]);
            var max = int.Parse(parts[0].Split("-")[1]);
            var letter = parts[1].Substring(0, 1);
            var password = parts[2];

            var count = countLetter1(password, letter);

            return count >= min && count <= max;
        }

        private static int countLetter1(string password, string letter)
        {
            var count = 0;
            var pos = password.IndexOf(letter);
            while (pos >= 0)
            {
                count++;
                pos = password.IndexOf(letter, pos+1);
            }

            return count;
        }

        public static int part2(IEnumerable<string> input) => input.Where(isValid2).Count();

        private static bool isValid2(string input)
        {
            var parts = input.Split(" ");
            var p1 = int.Parse(parts[0].Split("-")[0]);
            var p2 = int.Parse(parts[0].Split("-")[1]);
            var letter = parts[1][0];
            var password = parts[2];

            return password[p1 - 1] == letter ^ password[p2 - 1] == letter;
        }

    }
}
